### PR TITLE
(SLV-136) Create acceptance test for install operation

### PR DESCRIPTION
**Pull request by billclaytor**
This update adds the following acceptance tests for the ref_arch_setup install command:

1) Install on a remote target master by specifying the tarball URL
2) Install on a remote target master by specifying the tarball path on the controller
3) Install on a remote target master by specifying the tarball path already copied to the target master

TBD (these tests are still in development):
4) Install locally on the target master by specifying the tarball URL
5) Install locally on the target master by specifying the tarball path

The tests rely upon a pre-suite that configures the hosts (ssh, ruby), then copies the RAS gem and installs it on the controller host. The pre-suite also includes a test that downloads the tarball to both the controller and target-master (via Bolt task) for use in the "tarball path" scenarios.

Each test includes a teardown block that uninstalls puppet and removes the RAS temporary work directory. This allows the target_master host to be re-used for the three remote scenarios and for localhost scenarios that re-use the controller as a master (re-using a using a single rvm / ruby installation).

This update also includes a first pass at SLV-194 (RAS - update acceptance test to build gem locally). It currently builds the gem by calling the `gem:build` rake task from the `test:acceptance` task but does not yet provide the option to specify `RAS_GEM_PATH` or `RAS_GEM_VER`. This significantly reduces the acceptance test setup time by eliminating the need to scp the entire RAS directory.

### CODE REVIEW

File: acceptance/tests/30_install_test_remote_master_tarball_path_master.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    puts command
+    on controller, command
+  end
+
+  step "run puppet agent" do
+    on target_master, "puppet agent -t"
+  end
+
+  teardown do
+    # TODO: make this a task
```
Comment (RandellP) [inline acceptance/tests/30_install_test_remote_master_tarball_path_master.rb]: Move this teardown functionality to the beaker helper, and then call it from each test it is used in.

File: spec/ref_arch_setup/install_spec.rb [hunk 2]
```diff
@@ -541,6 +541,18 @@
         expect(install.copy_pe_tarball_on_target_master(pe_tarball)).to eq(false)
       end
     end
+
+    context "when the specified file is already in the TMP_WORK_DIR" do
+      it "returns false" do
```
Comment (RandellP) [inline spec/ref_arch_setup/install_spec.rb]: but line 553 says true. :)

File: Rakefile [hunk 3]
```diff
@@ -7,13 +7,83 @@ require "flay_task"
...
+# rubocop:disable Metrics/BlockLength
+namespace :test do
+  desc "Create hosts.cfg file"
+  task :mk_hosts_file do
+    beaker_initialize
+    beaker_create_host_file
+  end
+
+  desc "Run acceptance test using Beaker subcommands"
+  rototiller_task :acceptance do |_task|
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply Rakefile]: Updated to use a standard rake task without the |task| / |_task|

File: acceptance/config/options.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
+{
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply acceptance/config/options.rb]: Waiting for Sam's feedback

File: acceptance/helpers/beaker_helper.rb [hunk 5]
```diff
@@ -0,0 +1,208 @@
+# Beaker helper methods for use in running acceptance tests
+# rubocop:disable Metrics/ModuleLength
+module BeakerHelper
+  BEAKER_HOSTS = "#{__dir__}/../../hosts.cfg".freeze
+  BEAKER_KEYFILE = "".freeze
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply acceptance/helpers/beaker_helper.rb]: Removed

File: acceptance/helpers/beaker_helper.rb [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+  PE_TARBALL_EXTENSION = ".tar".freeze
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply acceptance/helpers/beaker_helper.rb]: Updated

File: acceptance/helpers/beaker_helper.rb [hunk 7]
```diff
@@ -0,0 +1,208 @@
+# Beaker helper methods for use in running acceptance tests
+# rubocop:disable Metrics/ModuleLength
+module BeakerHelper
+  BEAKER_HOSTS = "#{__dir__}/../../hosts.cfg".freeze
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply acceptance/helpers/beaker_helper.rb]: Updated

File: acceptance/pre_suites/30_setup_ras.rb [hunk 8]
```diff
@@ -0,0 +1,16 @@
+test_name "install RAS on controller" do
+  step "copy RAS gem, fixtures, and modules to controller" do
+    hosts = [controller]
+    scp_to(hosts, "#{__dir__}/../../pkg", "ref_arch_setup")
+    scp_to(hosts, "#{__dir__}/../../fixtures", "ref_arch_setup")
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply acceptance/pre_suites/30_setup_ras.rb]: RAS doesn't need them but the tests use the pe.conf to provide it to RAS, and the pre-suite uses the download_pe_tarball task to download the tarball for the "tarball path" scenarios. I can get the path to the installed gem and use the files there but found it easier to just copy them to the /root/ref_arch_setup folder and use them there.

File: acceptance/pre_suites/40_download_pe_tarball.rb [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    destination = "destination=/tmp/ras"
+    modulepath = "--modulepath /root/ref_arch_setup/modules"
+    nodes = "--nodes localhost,#{target_master}"
+    user = "--user root"
+
+    command = "#{bolt} #{task} #{url} #{destination} #{modulepath} #{nodes} #{user}"
+    puts command
+
+    on controller, command
+  end
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply acceptance/pre_suites/40_download_pe_tarball.rb]: The command and thus the step will fail if the download is not successful. This is a setup step providing the tarball that will be used in the "tarball path" scenarios, so the downloaded tarballs should not be removed.

File: acceptance/pre_suites/30_setup_ras.rb [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    scp_to(hosts, "#{__dir__}/../../modules", "ref_arch_setup")
+  end
+
+  step "install RAS on controller" do
+    hosts = [controller]
+    version = RefArchSetup::Version::STRING
+    gem = "ref_arch_setup-#{version}.gem"
+    command = "cd ref_arch_setup && gem install --local #{gem}"
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply acceptance/pre_suites/30_setup_ras.rb]: Updated to `command = "gem install $HOME/ref_arch_setup/#{gem}"`

File: spec/ref_arch_setup/install_spec.rb (hunk 2 above)
Comment (billclaytor) [reply spec/ref_arch_setup/install_spec.rb]: Updated

File: acceptance/pre_suites/30_setup_ras.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (RandellP) [reply acceptance/pre_suites/30_setup_ras.rb]: yeah, I would get them from the installed gem.  Should be a great deal faster.  Also, for RAS, if you don't pass in a pe.conf it should grab the one from the fixtures... which would be good for the test to try out as well.

File: acceptance/pre_suites/40_download_pe_tarball.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (RandellP) [reply acceptance/pre_suites/40_download_pe_tarball.rb]: yep, my bad, thought I was looking at a test, not a pre-suite

File: acceptance/tests/30_install_test_remote_master_tarball_path_master.rb (hunk 1 above)
Comment (billclaytor) [reply acceptance/tests/30_install_test_remote_master_tarball_path_master.rb]: Updated

File: acceptance/helpers/beaker_helper.rb [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+  # Initializes the PE instance variables
+  #
+  # @author Bill Claytor
+  #
+  # @return [void]
+  def beaker_initialize
+    @pe_family = ENV["BEAKER_PE_FAMILY"] || "2018.2"
+    @pe_url = "http://enterprise.delivery.puppetlabs.net/#{@pe_family}/ci-ready/LATEST"
+    curl_comm = "curl --silent #{@pe_url}"
+    @pe_version = ENV["BEAKER_PE_VERSION"] || `#{curl_comm}`.strip
```
[the remark this answers is not in the document]
Comment (RandellP) [reply acceptance/helpers/beaker_helper.rb]: actually, looking deeper... pe_family and pe_url are only used in this method, so they don't need to be @ vars.  Which solves the problem of them being out of sync.

[the remark this answers is not in the document]
Comment (billclaytor) [reply acceptance/helpers/beaker_helper.rb]: Updated

File: acceptance/helpers/beaker_helper.rb [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+
+  # Creates a Beaker host file for the acceptance tests
+  #
+  # @author Bill Claytor
+  #
+  # @return [exit_code] The result of the host file creation
+  #
+  def beaker_create_host_file
+    forge_host = ENV["BEAKER_FORGE_HOST"] || "forge-aio01-petest.puppetlabs.com"
+    hosts = "centos7-64controller.-64target_master."
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply acceptance/helpers/beaker_helper.rb]: Updated

File: acceptance/helpers/beaker_helper.rb [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+
+      if ENV.key?("BEAKER_PRE_SUITE")
+        command.add_option do |option|
+          option.name = "--pre-suite"
+          option.message = "Beaker pre-suite"
+          option.add_argument do |arg|
+            arg.add_env(name: "BEAKER_PRE_SUITE")
+          end
+        end
+
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply acceptance/helpers/beaker_helper.rb]: Updated

File: acceptance/helpers/beaker_helper.rb [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+      if ENV.key?("BEAKER_TESTS")
+        command.add_option do |option|
+          option.name = "--tests"
+          option.message = "Beaker tests"
+          option.add_argument do |arg|
+            arg.add_env(name: "BEAKER_TESTS")
+          end
+        end
+
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply acceptance/helpers/beaker_helper.rb]: Updated

File: acceptance/pre_suites/10_setup_ssh.rb [hunk 15]
```diff
@@ -0,0 +1,17 @@
+test_name "set up ssh between controller and target-master" do
+  step "create ssh key" do
+    on controller, "yes | ssh-keygen -q -t rsa -b 4096 -f /root/.ssh/id_rsa -N '' -C 'ras'"
+  end
+
+  step "put keys on the target-master" do
+    hosts = [target_master]
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply acceptance/pre_suites/10_setup_ssh.rb]: Added `hosts_without_role` to beaker_helper

**Pull request merged**